agent/local/state.go

```diff
@@ -325,10 +325,27 @@ func (l *State) RemoveServiceWithChecks(serviceID structs.ServiceID, checkIDs []
 	return nil
 }
 
+// Checks to see whether the given service ID matches the name of an existing
+// service. Used to see whether a service lookup failed because the user
+// mistakenly provided a service name rather than a service ID.
+func (l *State) doesServiceExistWithNameOfID(lookupID structs.ServiceID) bool {
+	for _, searchTarget := range l.services {
+		if searchTarget.Service.Service == lookupID.ID {
+			return true
+		}
+	}
+
+	return false
+}
+
 func (l *State) removeServiceLocked(id structs.ServiceID) error {
 	s := l.services[id]
 	if s == nil || s.Deleted {
-		return fmt.Errorf("Service %q does not exist", id)
```

Comment:
Could we simply improve the error message to be something like:
```
"Service %q does not exist. Ensure that the service ID is passed, not the service name."
```

I'm a _little_ bit wary of the code doing extra `O(l.services)` work just to pass a better error message (in which case we might as well output the found service ID anyways).

Reply:
@kisunji : my understanding (which could be wrong) is that service ID is required to be unique, but service name is not. I think that's why the endpoint requires a service ID to be passed rather than a name in the first place.

How often does the deregister endpoint get called/in what context? If it doesn't happen often, as long as `l.services` isn't that large, does the extra work have a meaningful impact?

If we are worried about that extra `O(l.services)` work, I like your suggestion to alter the error message in every case.

Reply:
The extra work is probably negligible, but I'd prefer a general error message for "not found" over splitting error states to "not found" and "not found but service with that name found". Just my opinion though! Could try to get more eyes from the team.

Reply:
How about 
```
"Service with ID %q does not exist. Ensure that the service ID is passed, not the service name."
```

My thought is that a customer who sees the new error message may think "well, if you know the service name, just deregister it!".  I understand your point about uniqueness for ID vs. name, but the customer probably doesn't realize that and may end up more frustrated.

Reply:
I didn't originally go in that direction because I didn't want to mislead someone who was using the endpoint correctly, just with an ID for a service that does not exist anymore (or never did). Perhaps there's a way of altering the wording to satisfy both cases.

More generally, long term, it would be nice for error messages to include a unique ID that allows a user to easily look up troubleshooting information so that the nuance/steps can be conveyed there (while still keeping the essentials in the error message). Not relevant right now for this PR, though.

Reply:
> The extra work is probably negligible, but I'd prefer a general error message for "not found" over splitting error states to "not found" and "not found but service with that name found".

I think it's worth getting other opinions so that we can have a consistent decision across future, similar situations. I prefer giving users as specific a message as we can for what they did wrong, though I'm open to being persuaded otherwise!

Reply:
I don't find the time complexity *too* objectionable, but I do have one question.

How does this interact with authorization? Do we check auth for the name? Are we potentially providing an oracle that could allow users to bypass auth and discover names?

Reply:
@msiege2: Regarding `Service with ID %q does not exist.` ... I had been thinking the same thing. That does beg the question though, should all messages in that file be updated from `service %q` to `service with ID %q`?

e.g.,
`"cannot add service %q to node in partition %q"`
`"Check %q refers to non-existent service %q"`

Reply:
@markan:

Those are good questions which raise yet another ...

From my read of the endpoint function `AgentDeregisterService`, the `RemoveService` call happens after `vetServiceUpdateWithAuthorizer`. Because `vetServiceUpdateWithAuthorizer` checks to see if a service with that ID exists, and returns an error if there is no service with that ID found, I'm not sure how the `RemoveService` call would ever even be reached from the endpoint. I would need to test this differently to ensure this error message is generated in response to incorrect use of the endpoint (which was the original goal).

https://github.com/hashicorp/consul/blob/d10dd5fbb196934560d72b4d48c723cc5481cb0c/agent/agent_endpoint.go#L1135-L1163

https://github.com/hashicorp/consul/blob/d10dd5fbb196934560d72b4d48c723cc5481cb0c/agent/acl.go#L73-L88

Reply:
I agree with @kisunji. Adding logic to check if the user is mistakenly providing a service name instead of an ID to only change the error message is adding extra code complexity that is not justified in my opinion.

I think it's better if we just reword the generic error message to specify ID `Service with ID %q does not exist` for example.

Reply:
I ran a Consul dev instance with one service registered (name: `counting`, id: `counting-id`). When I called `/agent/service/deregister` with values `test`, `counting`, and `counting-id`, here's what appeared in the Consul logs:

```
2021-12-09T13:19:39.127-0800 [ERROR] agent.http: Request error: method=PUT url=/v1/agent/service/deregister/test from=127.0.0.1:45106 error="Unknown service "test""
2021-12-09T13:20:01.860-0800 [ERROR] agent.http: Request error: method=PUT url=/v1/agent/service/deregister/counting from=127.0.0.1:45108 error="Unknown service "counting""
2021-12-09T13:20:09.564-0800 [INFO]  agent: Deregistered service: service=counting-id
```

This suggests that, at least with Consul 1.10+, the error message I changed here isn't even exposed to the end user. The error message exposed to the user is definitely coming from `vetServiceUpdateWithAuthorizer`, which returns an error if a service doesn't exist before `RemoveService` is ever called.

https://github.com/hashicorp/consul/blob/bcce800f5514cd99b1d4173051ad9f8cfd8606fa/agent/acl.go#L86

Reply:
Ah yeah I've been bit with that as well. I don't see any reason why we can't change the error message to be a bit more descriptive in `vetServiceUpdateWithAuthorizer` (as well as many other places where we return `"Unknown <entity> %q"`

Reply:
I've made the update in `vetServiceUpdateWithAuthorizer`. I'll see what else I can find of the form `"Unknown <entity> %q"` (e.g., health checks).

Reply:
It looks like there's actually source code that depends on the string value of these error messages. I've found one so far. I'm worried there might be others.

https://github.com/hashicorp/consul/blob/bcce800f5514cd99b1d4173051ad9f8cfd8606fa/agent/local/state.go#L1259

Perhaps to reduce risk I should make the error message change to
```
"Unknown service with ID %q. Ensure that the service ID is passed, not the service name."
```

instead of:

```
"Service with ID %q does not exist. Ensure that the service ID is passed, not the service name."
```

?

Reply:
The if-conditional above actually makes me wonder why the old error message (`"Service %q does not exist"`) never caused a problem, because it doesn't contain `"Unknown service"`... Maybe only because that error message was always (or in most cases) masked by `vetServiceUpdateWithAuthorizer`.

Reply:
Yeah if the test setup didn't involve ACLs it wouldn't have gotten that error